### PR TITLE
TempRValueOptPass: accept try_apply as user of the temporary object

### PR DESCRIPTION
**Pull request by eeckstein**
So far we only handled apply, but not try_apply.

Also: some small cosmetic changes

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILOptimizer/Transforms/CopyForwarding.cpp [hunk 1]
```diff
@@ -1645,7 +1645,8 @@ bool TempRValueOptPass::collectLoads(
                             << *user);
     return false;
 
-  case SILInstructionKind::ApplyInst: {
+  case SILInstructionKind::ApplyInst:
+  case SILInstructionKind::TryApplyInst: {
     ApplySite apply(user);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CopyForwarding.cpp]: Can this ever work if something isn't a full apply site? If not, can you make this a full apply site?

File: lib/SILOptimizer/Transforms/CopyForwarding.cpp [hunk 2]
```diff
@@ -1714,12 +1715,10 @@ bool TempRValueOptPass::collectLoads(
 
   case SILInstructionKind::LoadInst:
   case SILInstructionKind::LoadBorrowInst:
-  case SILInstructionKind::WitnessMethodInst: {
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CopyForwarding.cpp]: Why did you remove the rest of this stuff from this PR?

Comment (eeckstein) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: I added more detailed comment in the commit message (I should have copied that to the PR description):

1. WitnessMethodInst doesn't need to be handled as use because it's operand is a typedependent operand, which we exclude anyway when looking at the uses.
2. Replace a list of handled instructions with a default value in the switch. It's easy to forget to add an instruction here. So it's safer to have a default behavior.

I hope that answers your question

File: lib/SILOptimizer/Transforms/CopyForwarding.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: Yeah, that would probably be more descriptive. But using an ApplySite does not hurt here, because the switch case already selects only full apply sites: ApplyInst and TryApplyInst.

File: lib/SILOptimizer/Transforms/CopyForwarding.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: I think this is the wrong approach. It will make it so that we by mistake handle things that we may not intend to. That is a large problem with this pass.

Thought: why not instead make it so that the default in asserts builds asserts, but in non-asserts builds has the other behavior?

Comment (eeckstein) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: I don't agree. We want to handle a few instructions in a special way, all other instructions the same. So IMO, using "default" is the right choice here.

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: The statement "we want to handle a few instruction in a special way, all other instructions the same" is incorrect.

When this pass was written we were pattern matching a specific semantic codegen pattern, so the code was written this way by @atrick to make sure that we /only/ pattern matched that pattern and if we hit a pattern by mistake that was not anticipated, we got a nice error from the compiler so we knew to update this code.

With this change, we will not get that error and instead will start silently accepting such new code patterns, allowing for the introduction of bugs and a reduction in the quality and correctness of the compiler.

Comment (atrick) [reply lib/SILOptimizer/Transforms/CopyForwarding.cpp]: I don't really care about this particular case--it seems fairly hard to break. But Michael is right that the original intention was for the optimization to be self-checking.

When correctness is at stake, it's generally important to list both the specially handled instructions and the "known-safe" instructions. There are so many variations of SIL operations, and we're continually adding more, so new instructions might come along that need special handling. (The previous bug was that an instruction *was* specially handled, but handled incorrectly).

Listing known-safe cases also communicates to future maintainers which cases were deliberately handled vs. which cases were forgotten. It helps me when reading the code to know what all the expected cases are without taking time to step through the code.

There is always the tradeoff between aggressively asserting in optimization passes vs. falling back to something that is probably safe. The LLVM/Swift approach has generally been to aggressively assert. It means many more bug reports early on but fewer hard-to-reproduce bugs later in the produce lifecycle.

**Review by atrick — COMMENTED** [review]
LGTM: I don't know why we weren't handling try_apply before.
